### PR TITLE
Update all the runtimes

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Closes https://github.com/crossplane/crossplane/pull/4223

Namely crossplane-runtime, and controller-runtime. Also k8schain.

A few breaking changes in this new controller-runtime. We need to bump to be able to bump to k8s v0.27.x, which other dependencies like k8schain are starting to require.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR if necessary.~

### How has this code been tested

[contribution process]: https://git.io/fj2m9

E2E tests are passing. 😄

### CODE REVIEW

File: cmd/crossplane/core/core.go
Comment (phisco) [inline cmd/crossplane/core/core.go:241]: Does this mean we are going to deploy webhooks always now? (I have to check the code better, just wanted to drop a comment before forgetting and before this got merged)

Comment (negz) [reply cmd/crossplane/core/core.go:241]: Yeah this is something I want to understand better too. I _think_ before this change:

* A webhook server was always started and listening (by controller runtime).
* If `c.WebhookTLSCertDir` was supplied, we'd tweak its config and register our webhooks.
* We'd only actually expose the webhook server if `webhooks.enabled` was true in the Helm chart

My thinking with this change was it might be fine to just register the webhooks whether they were going to be exposed/used or not.

Comment (negz) [reply cmd/crossplane/core/core.go:241]: Hrm, no this now breaks. I get the following when I set `webhooks.enabled=false`:

```
crossplane: error: core.startCommand.Run(): Cannot start controller manager: open /tmp/k8s-webhook-server/serving-certs/tls.crt: no such file or directory
```

Comment (phisco) [reply cmd/crossplane/core/core.go:241]: yes, that was what I was expecting, due to how webhooks are handled we also hit this issue: https://github.com/crossplane/crossplane/issues/4095
I don't think we should address it here, but could be worth keeping it in mind and if it becomes to complex we could think of alternatives, e.g. moving validatingwebhookconfiguration creation to the helm chart and leave in the controller only the ca injection

Comment (negz) [reply cmd/crossplane/core/core.go:241]: I tweaked things a bit (see new commits) but it's still not passing the E2E suite in https://github.com/crossplane/crossplane/pull/4187.

Comment (negz) [reply cmd/crossplane/core/core.go:241]: Looks like a lot of breaking changes related to webhooks in controller-runtime v0.15. I've added a commit that implements an `admission.CustomValidator`, but I'm still seeing failures when I run the E2E tests. Interestingly the failures are now to do with XRDs. 🤔 

```
2023-06-22T09:10:09Z    ERROR   crossplane      Reconciler error        {"controller": "defined/compositeresourcedefinition.apiextensions.crossplane.io", "controllerGroup": "apiextensions.crossplane.io", "controllerKind": "CompositeResourceDefinition", "CompositeResourceDefinition": {"name":"xnopresources.nop.example.org"}, "namespace": "", "name": "xnopresources.nop.example.org", "reconcileID": "3defa3c9-085d-4b2f-9f77-b88eb7322ec3", "error": "cannot add composite resource finalizer: cannot update object: Internal error occurred: failed calling webhook \"compositeresourcedefinitions.apiextensions.crossplane.io\": failed to call webhook: the server could not find the requested resource"}
```

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (negz) [inline internal/validation/apiextensions/v1/composition/handler.go:61]: This no longer works with controller-runtime v0.15.

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (negz) [inline internal/validation/apiextensions/v1/composition/handler.go:80]: I removed the `unstructured.NewClient` that wraps the manager's client here. I can't see any reason we'd need it. Let me know if I missed it.

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (negz) [inline internal/validation/apiextensions/v1/composition/handler.go:61]: @phisco Do you know. I can kinda guess but it was confusing enough that a comment would be useful.

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (phisco) [reply internal/validation/apiextensions/v1/composition/handler.go:61]: yes, this and the whole implementation of the Handler can be removed, it was only needed because we weren't able to use the CustomValidator until https://github.com/kubernetes-sigs/controller-runtime/issues/1896 was merged and released as said a few line below.

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (phisco) [reply internal/validation/apiextensions/v1/composition/handler.go:61]: Something like this would make sense?
```suggestion
		// set up indexes to be able to list CRDs by Group and Kind they define later on
```

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (phisco) [reply internal/validation/apiextensions/v1/composition/handler.go:80]: I think it was needed for rendered resource validation IIRC, fine removing it now, I'll have to manually test it though, would be nice to have an e2e for this.

File: cmd/crossplane/core/core.go
Comment (negz) [reply cmd/crossplane/core/core.go:241]: Found it - XRD needed its methods updated to be a valid `admission.Validator`. Pushed a commit.

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (negz) [reply internal/validation/apiextensions/v1/composition/handler.go:80]: @phisco I added a rudimentary E2E test, which seems to pass. Can you take a look and confirm whether it's actually testing anything useful? Maybe I should enable strict validation on the relevant `Composition`.

File: test/e2e/apiextensions_test.go
Comment (negz) [inline test/e2e/apiextensions_test.go:199]: I wonder whether it would make sense to have every test "upgrade" the Crossplane deployment to have whatever flags they need. Kinda weird, but it could be more resilient than relying on tests that change how Crossplane is installed to remember to undo their changes.

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (phisco) [reply internal/validation/apiextensions/v1/composition/handler.go:80]: The test case looks good to me as an initial smoke test, I would maybe just add another step checking that an invalid composition is rejected, e.g.:
Change the XRD to have one more `evenCoolerField` of type integer:
```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata:
  name: xnopresources.nop.example.org
spec:
  group: nop.example.org
  names:
    kind: XNopResource
    plural: xnopresources
  claimNames:
    kind: NopResource
    plural: nopresources
  connectionSecretKeys:
  - test
  versions:
  - name: v1alpha1
    served: true
    referenceable: true
    schema:
     openAPIV3Schema:
       type: object
       properties:
        spec:
          type: object
          properties:
            coolField:
              type: string
          required:
          - coolField
        status:
          type: object
          properties:
            coolerField:
              type: string
            evenCoolerField:               # <== added field
              type: integer
```
which then we are going to patch from a string without any transformation:
```yaml
---
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: xnopresources.nop.example.org
spec:
  compositeTypeRef:
    apiVersion: nop.example.org/v1alpha1
    kind: XNopResource
  resources:
  - name: nop-resource-1
    base:
     apiVersion: nop.crossplane.io/v1alpha1
     kind: NopResource
     spec:
      forProvider:
        conditionAfter:
        - conditionType: Ready
          conditionStatus: "False"
          time: 0s
        - conditionType: Ready
          conditionStatus: "True"
          time: 10s
    patches:
    - type: FromCompositeFieldPath
      fromFieldPath: spec.coolField
      toFieldPath: metadata.annotations[cool-field]
      transforms:
      - type: string
        string:
         type: Convert
         convert: ToUpper
    - type: ToCompositeFieldPath
      fromFieldPath: metadata.annotations[cool-field]
      toFieldPath: status.coolerField
    - type: ToCompositeFieldPath                                # <== added patch
      fromFieldPath: metadata.annotations[cool-field]
      toFieldPath: status.evenCoolerField
```
And once applied we should get an error, I wouldn't check here the specific error, as that's checked by the unit tests, but manually applying it results in the following error:
```
The Composition "xnopresources.nop.example.org" is invalid: spec.resources[0].patches[2].transforms: Required value: the fromFieldPath does not have a type compatible with the toFieldPath according to their schemas and no transforms were provided: string != integer
```

File: apis/apiextensions/v1/xrd_webhooks.go
Comment (negz) [reply apis/apiextensions/v1/xrd_webhooks.go:41]: Let me know if there's an action for me to take here.

File: apis/apiextensions/v1/xrd_webhooks.go
Comment (negz) [reply apis/apiextensions/v1/xrd_webhooks.go:48]: Let me know if there's an action for me to take here.

File: apis/apiextensions/v1/xrd_webhooks.go
Comment (negz) [reply apis/apiextensions/v1/xrd_webhooks.go:86]: > Do you expect to add here the code to setup al the webhooks for all resources in this package in one go?

Eventually - that's would be more similar to what we do with controllers (per the TODO in `core.go`).

Largely my thinking here was that where it's called:

```go
apiextensionsv1.SetupWebhookWithManager(mgr)
```

Seems a lot simpler (and perhaps more idiomatic?) than the approach where this is a method on a type:

```go
(&apiextensionsv1.CompositeResourceDefinition{}).SetupWebhookWithManager(mgr)
```

To me, it's a little confusing/unintuitive that calling Setup on one (empty) instance of a type is registering the entire type with the controller manager. Admittedly that's still happening with my update - it's just more in line with how we do every other `For(T)` call in the codebase, and IMO reads more like "setup a webook for this type" rather than "setup a webhook for this instance of this type".

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (negz) [reply internal/validation/apiextensions/v1/composition/handler.go:54]: That's good to know. In cases like this the best thing to do is to raise an issue to discuss changing the pattern. Otherwise we start to have some files that are look like @phisco wrote them according to his preferences, others like @negz wrote them etc. 🙂 Would you mind raising an issue to discuss?

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (negz) [reply internal/validation/apiextensions/v1/composition/handler.go:115]: That seems like a good idea. I think first step would be to surface that function in `pkg/errors` in crossplane-runtime. I'll update the TODO.

I can see in this case where we're validating an object that it's more worth it than usual to try process the whole thing and return all errors so that you don't have to iterate through each error case to discover them.

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (negz) [reply internal/validation/apiextensions/v1/composition/handler.go:122]: Agreed - wrapped that mistakenly.

File: test/e2e/apiextensions_test.go
Comment (negz) [reply test/e2e/apiextensions_test.go:145]: I don't think I understand what you mean. That said, I don't want to change the established pattern in this PR. Could you raise a new issue (and label it `e2e`) to discuss please.

File: test/e2e/apiextensions_test.go
Comment (negz) [reply test/e2e/apiextensions_test.go:199]: I'm going to keep this as is - we can change if/when we find a better pattern.

File: apis/apiextensions/v1/xrd_webhooks.go
Comment (phisco) [reply apis/apiextensions/v1/xrd_webhooks.go:48]: If you want to enable it by adding the create verb to the annotation above, otherwise I can do it on a separate PR next week.

File: apis/apiextensions/v1/xrd_webhooks.go
Comment (phisco) [reply apis/apiextensions/v1/xrd_webhooks.go:86]: I agree it's not too nice, but I got used to it probably, it's a pretty common pattern: https://grep.app/search?q=Setupwebhookwithmanager

File: internal/validation/apiextensions/v1/composition/handler.go
Comment (negz) [reply internal/validation/apiextensions/v1/composition/handler.go:80]: Done!

File: apis/apiextensions/v1/xrd_webhooks.go
Comment (negz) [reply apis/apiextensions/v1/xrd_webhooks.go:48]: That seems like it could be a functional change, so let's keep it in a separate PR.

File: apis/apiextensions/v1/xrd_webhooks.go
Comment (negz) [reply apis/apiextensions/v1/xrd_webhooks.go:86]: Yeah it's not fundamentally wrong, and is what controller-runtime scaffolds I think. We diverged from that scaffolding with our `Setup` (for controller) pattern so I'd like to keep consistent.

**Comment by negz** [comment]
Adding a note that we now get pprof for free with controller-runtime, so I'd like to remove the one I manually setup. Doesn't need to block this PR though - can follow up later. https://github.com/kubernetes-sigs/controller-runtime/pull/1943

File: internal/validation/apiextensions/v1/composition/handler_test.go
Comment (phisco) [inline internal/validation/apiextensions/v1/composition/handler_test.go:21]: Why the need to have this in a test file instead of directly in internal/validation/apiextensions/v1/composition/handler.go?

**Review by phisco — APPROVED** [review]

File: internal/validation/apiextensions/v1/composition/handler_test.go
Comment (negz) [reply internal/validation/apiextensions/v1/composition/handler_test.go:21]: Because it's a test. 🙂 We're testing that our `*validator` type satisfies the interface we need it to satisfy. No need for this code to be part of our production binary.

**Pull request merged**
